Implement `create_psbt` for Wallet

ValuedMammal opened the pull request:
### Description

This PR introduces a new PSBT construction flow and associated wallet APIs.

Primary additions:

- New `PsbtParams<Ctx>` API with typed contexts (`CreateTx` and `ReplaceTx`) for transaction creation and replacement.
- New wallet entry points for PSBT construction and RBF:
	- `Wallet::create_psbt` and `Wallet::create_psbt_with_rng`
	- `Wallet::replace_by_fee`, `Wallet::replace_by_fee_with_rng`, and `Wallet::replace_by_fee_and_recipients`
- New error types for the new create-PSBT and RBF paths.
- Supporting wallet helpers for transaction insertion and canonicalization-aware transaction listing.
- New examples and expanded test coverage for expected PSBT/RBF behavior.
- Generic `TxOrdering<In, Out>` updates used by the new selection/PSBT path

This work is part of the broader migration of transaction construction to `bdk_tx`, which uses miniscript's `plan` module under the hood. A key long-term goal is the move away from legacy policy-based transaction construction internals (#4).

fix #204 

### Notes to the reviewers

Suggested review order for this PR:

1. Start with `examples/psbt.rs` and `examples/replace_by_fee.rs` for a high-level API walkthrough.
2. Review `src/psbt/params.rs` in detail (types, state transitions, params semantics, and so on).
3. Review the new wallet implementation in `src/wallet/mod.rs` (create-PSBT and RBF execution paths plus shared helpers).
4. Review tests in `tests/psbt.rs`, `tests/wallet.rs` to ensure expected behavior and edge cases are covered.


### Changelog notice

### Changed

- `TxOrdering::Custom` is generic over the input and output types.

### Added

- `psbt::params` module
- `PsbtParams` struct
- `CoinSelectionStrategy` enum
- `UtxoFilter` struct

Added `pub` methods on `Wallet`

- `Wallet::create_psbt{_with_rng}`
- `Wallet::replace_by_fee{_with_rng}`
- `Wallet::replace_by_fee_and_recipients`

Added examples
- `examples/psbt.rs`
- `examples/replace_by_fee.rs`

### Checklists

* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [ ] ~~This pull request breaks the existing API~~

thunderbiscuit reviewed (commented):
So this is more of a conceptual review/question set, as I'm getting acquainted with the PR, and it also requires knowledge and understanding of the bdk-tx crate/workflow.

One general question I have is do you think is missing in functionality between this and the current TxBuilder? Can we make a todo list that compares functionality with the current TxBuilder to better visualize how close of a replacement this is, or if it only provides part of the functionality for now (and if so which parts)?

I haven't had time to look/test the examples and my day is over, but I'll come back to this on Monday.

oleonardolima commented on wallet/Cargo.toml:
```diff
@@ -22,12 +22,14 @@ bitcoin = { version = "0.32.6", features = [ "serde", "base64" ], default-featur
 bip39 = { version = "2.0", optional = true }
 bdk_file_store = { version = "0.21.1", optional = true }
+anyhow = { version = "1.0.98", optional = true }
+tempfile = { version = "3.20.0", optional = true }
```
I guess this can be removed, as it's not being used.

ValuedMammal replied:
These are currently being used by the `test-utils` feature, but there could be a way to convert them back to dev dependencies.

oleonardolima commented on wallet/src/wallet/error.rs:
```diff
@@ -256,3 +256,68 @@ impl fmt::Display for BuildFeeBumpError {
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum CreatePsbtError {
+    /// No Bnb solution.
+    Bnb(bdk_coin_select::NoBnbSolution),
+    /// Non-sufficient funds
+    InsufficientFunds(bdk_coin_select::InsufficientFunds),
```
Can't we have a generic CS error instead ?

oleonardolima commented on wallet/src/wallet/mod.rs:
````diff
@@ -2635,6 +2687,495 @@ impl Wallet {
+    /// # Ok::<_, anyhow::Error>(())
+    /// ```
+    #[cfg(feature = "std")]
+    pub fn create_psbt(&self, params: PsbtParams) -> Result<(Psbt, Finalizer), CreatePsbtError> {
````
I still need to take a deeper look into bdk-tx, but I think the name could be misleading with `Input Finalizer` from the PSBT BIP-174 (if it's not following the same proposed idea).

oleonardolima reviewed (commented):
I did an initial review, with simple comments/questions. Still need to take a deeper look into bdk-tx and do a more thorough review here.

ValuedMammal commented on wallet/src/wallet/mod.rs:
```diff
@@ -2635,6 +2687,495 @@ impl Wallet {
+    ///
+    /// Return `None` if `outpoint` doesn't correspond to an indexed txout, or
+    /// if the assets are not sufficient to create a plan.
+    fn try_plan(&self, outpoint: OutPoint, assets: &Assets) -> Option<Plan> {
```
https://hackmd.io/@valuedmammal/B1lxzomj0

ValuedMammal commented on wallet/src/psbt/params.rs:
```diff
@@ -0,0 +1,335 @@
+    ///
+    /// A single outpoint may appear at most once in the list of UTXOs to spend. The caller is
+    /// responsible for ensuring that elements of `outpoints` correspond to outputs of previous
+    /// transactions and are currently unspent.
```
🤔

ValuedMammal commented on wallet/src/psbt/params.rs:
```diff
@@ -0,0 +1,335 @@
+/// `ReplaceParams` provides a thin wrapper around [`PsbtParams`] and is intended for
+/// crafting Replace-By-Fee transactions (RBF).
+#[derive(Debug, Default)]
+pub struct ReplaceParams {
```
> Naively I would have thought you'd just call `PsbtParams::replace` and you'd get a sort of pre-populated PsbtParams ready for replacing your tx

That works too.

thunderbiscuit replied:
If that works that's my preferred approach, unless there is something I'm not seeing.

ValuedMammal replied:
Removed `ReplaceParams`. Now calling `PsbtParams::replace_txs` will return a new `PsbtParams` that is marked by the "RBF" context. I think it's a cleaner approach overall without compromising type safety.

thunderbiscuit commented on wallet/src/psbt/params.rs:
```diff
@@ -0,0 +1,335 @@
+    }
+
+    /// Set the definite descriptor used for generating the change output.
+    pub fn change_descriptor(&mut self, desc: DefiniteDescriptor) -> &mut Self {
```
Ok awesome thanks for confirming. In this case, I'd just mention this in the docs (that this method is basically "if you want to send change elsewhere than your default change location)"

ValuedMammal commented:
### TxBuilder vs PsbtParams feature comparison

*subject to change*

| Feature                              | TxBuilder  | PsbtParams | Status    |
| ---------                            | ---------- | ---------- |---------- |
| internal_policy_path                 |    ✅      |     ❌     | deprecated  |
| external_policy_path                 |    ✅      |     ❌     | deprecated  |
| manually_selected_only               |    ✅      |     ❌     | unknown    |
| include_output_redeem_witness_script |    ✅      |     ❌     | deprecated |
| current_height                       |    ✅      |     ❌     |  unknown** |
| fee_policy                           |    ✅      |    ✅ ❌   |  TODO |
| sighash (type)                       |    ✅      |     ❌     |  TODO |
| ordering                             |    ✅      |     ❌     |  TODO |
| add_global_xpubs                     |    ✅      |     ❌     |  TODO |
| allow_dust                           |    ✅      |     ❌     |  unknown** |
| unspendable                          |    ✅      |     ❌     |  unknown** |
| recipients                           |    ✅      |     ✅     |  Done |
| utxos                                |    ✅      |     ✅     |  Done |
| drain_wallet                         |    ✅      |     ✅     |  Done |
| drain_to/change_script               |    ✅      |     ✅     |  Done |
| locktime                             |    ✅      |     ✅     |  Done |
| version                              |    ✅      |     ✅     |  Done |
| fallback_sequence                    |    ✅      |     ✅     |  Done |
| bumping_fee                          |    ✅      |     ✅     |  Done* |
| only_witness_utxo                    |    ✅      |    ❌      |  unknown** |
| change_policy                        |    ✅      |    ❌      |  TODO |
| utxo_filter                          |    ❌      |    ❌      |  TODO |

\* See RbfParams
\** or undecided

ValuedMammal commented:
- We need a way to implement `fee_absolute` or find out if this is possible using `bdk_coin_select`.
- Note that `UtxoFilter` can accomplish a number of things from the old interface, such as "change policy" and "unspendable".

ValuedMammal commented:
> We need a way to implement `fee_absolute` or find out if this is possible using `bdk_coin_select`.

I think a straightforward approach is to set a target feerate of `0` and add the fee amount to the output [`value_sum`](https://docs.rs/bdk_coin_select/latest/bdk_coin_select/struct.TargetOutputs.html#structfield.value_sum). The effect is to raise the cost of meeting the target by the amount of the target fee.

I can see it being relevant to CPFP, since there we're less concerned about the feerate of an individual tx and more so with the fee needed to bump a package to a target feerate.

DanGould commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,693 @@
+    /// Optional assets for creating a spend plan.
+    pub(crate) assets: Option<Assets>,
+    /// Fee targeting strategy.
+    pub(crate) fee_strategy: FeeStrategy,
```
This is what our API looks like to solve our fee problems for receivers: https://docs.rs/payjoin/1.0.0-rc.1/payjoin/receive/v2/struct.Receiver.html#impl-Receiver%3CWantsFeeRange%3E

Internally we enforce the minimum absolute fee and we enable a receiver to specify their minimum fee rate when the initialize the receiver (though I suppose in time it should really be a function rather than a value) as well as a maximum fee rate to prevent overspending by a counterparty who insists on a higher feerate.

very much a nit, but '"ensures" the [transaction] gets received in the next block' is necessarily interactive, and still only a best effort, since all we can do is estimate in this function afaiu.

ValuedMammal commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,697 @@
+    ///
+    /// If adding a time-based absolute timelock here, you must also set a [`PsbtParams::locktime`]
+    /// which is time-based (i.e. UNIX timestamp) to use as a fallback, otherwise a
+    /// `LockTypeMismatch` error will occur.
```
I find this slightly annoying and there could be an alternative way to handle it. Edit: Fixed by bitcoindevkit/bdk-tx#43.

noahjoeris commented:
One thing I noticed is that bdk-tx test coverage is a bit thin right? Might be good to add some more before merging. I’ll have a look and see if I can help :)

ValuedMammal commented:
- Changed the `examples/replace_by_fee.rs` to a more practical example of a sweep transaction that gets replaced with a higher fee

ValuedMammal committed:
feat(psbt): Add PsbtParams

We add the `psbt::params` module along with new types
including `PsbtParams` and `SelectionStrategy`.

`PsbtParams` is mostly inspired by `TxParams` from `tx_builder.rs`,
except that we've removed support for `policy_path` in favor of
`add_assets` API.

`PsbtParams<C>` contains a type parameter `C` indicating the context
in which the parameters can be used. Methods related to PSBT
creation exist within the `CreateTx` context, and methods
related to replacements (RBF) exist within the `ReplaceTx`
context.

In `lib.rs` re-export everything under `psbt` module.

- deps: Add git dependency `bdk_tx` 0.2.0

ValuedMammal committed:
feat(psbt): Add PsbtParams

We add the `psbt::params` module along with new types
including `PsbtParams` and `SelectionStrategy`.

`PsbtParams` is mostly inspired by `TxParams` from `tx_builder.rs`,
except that we've removed support for `policy_path` in favor of
`add_assets` API.

`PsbtParams<C>` contains a type parameter `C` indicating the context
in which the parameters can be used. Methods related to PSBT
creation exist within the `CreateTx` context, and methods
related to replacements (RBF) exist within the `ReplaceTx`
context.

In `lib.rs` re-export everything under `psbt` module.

- deps: Add `bdk_tx` 0.2.0 to Cargo.toml

ValuedMammal committed:
feat(tx_builder): Generalize TxOrdering inputs, outputs

`TxOrdering` is made generic by exposing the generic from
`TxSort` function. This means we're not limited to
ordering lists of only `TxIn` and `TxOut`, which will be
useful for sorting inputs/outputs of a `bdk_tx::Selection`.

We use bitcoin `TxIn` and `TxOut` as the default type parameter
to maintain backward compatibility.

Add `sort_with_rng` for `TxOrdering<In, Out>` for sorting
two generic mutable slices.

ValuedMammal committed:
feat(tx_builder): Generalize TxOrdering inputs, outputs

`TxOrdering` is made generic by exposing the generic from
`TxSort` function. This means we're not limited to
ordering lists of only `TxIn` and `TxOut`, which will be
useful for sorting inputs/outputs of a `bdk_tx::Selection`.

We use bitcoin `TxIn` and `TxOut` as the default type parameter
to maintain backward compatibility.

ValuedMammal committed:
feat(wallet): Add `Wallet::create_psbt`

We use the new `PsbtParams` to add methods on `Wallet` for
creating PSBTs, including RBF transactions.

`Wallet::create_psbt` and `Wallet::replace_by_fee` each have
no-std counterparts that take an additional `impl RngCore`
parameter.

Also adds a convenience method `replace_by_fee_and_recipients`
that exposes the minimum information needed to create an RBF.

This commit re-introduces the `Wallet::insert_tx` API for adding
newly created transactions to the wallet.

Added `Wallet::transactions_with_params` that allows customizing
the internal canonicalization logic.

Added errors to `wallet::errors` module:
- `CreatePsbtError`
- `ReplaceByFeeError`

ValuedMammal committed:
feat(wallet): Add `Wallet::create_psbt`

We use the new `PsbtParams` to add methods on `Wallet` for
creating PSBTs, including RBF transactions.

`Wallet::create_psbt` and `Wallet::replace_by_fee` each have
no-std counterparts that take an additional `impl RngCore`
parameter.

Also adds a convenience method `replace_by_fee_and_recipients`
that exposes the minimum information needed to create an RBF.

This commit re-introduces the `Wallet::insert_tx` API for adding
newly created transactions to the wallet.

Added `Wallet::transactions_with_params` that allows customizing
the internal canonicalization logic.

Added errors to `wallet::errors` module:
- `CreatePsbtError`
- `ReplaceByFeeError`

ValuedMammal committed:
test: Add coverage for `create_psbt` and related fns

Added unit test to `psbt/params.rs`
- `test_replace_params`

To tests/add_foreign_utxo.rs added
- `test_add_planned_psbt_input`

To tests/psbt.rs added
- `test_create_psbt`
- `test_create_psbt_insufficient_funds_error`
- `test_create_psbt_maturity_height`
- `test_create_psbt_cltv`
- `test_create_psbt_cltv_timestamp`
- `test_create_psbt_csv`
- `test_replace_by_fee_and_recpients`
- `test_replace_by_fee_replaces_descendant_fees`
- `test_replace_by_fee_confirmed_tx_error`
- `test_replace_by_fee_no_inputs_from_original`
- `test_create_psbt_utxo_filter`

Plus several Sequence fallback and override scenarios
- `test_create_psbt_fallback_sequence_applied_to_coin_selected_input`
- `test_create_psbt_fallback_sequence_skipped_for_csv_input`
- `test_create_psbt_sequence_override_manually_selected_input`
- `test_create_psbt_sequence_override_takes_precedence_over_fallback`
- `test_create_psbt_sequence_override_csv_conflict_returns_error`

To tests/wallet.rs added
- `test_spend_non_canonical_txout`

- `test-utils`: Add `insert_tx_anchor` test helper for adding
a transaction to the wallet with associated anchor block.

ValuedMammal committed:
test: Add coverage for `create_psbt` and related fns

Added unit test `psbt::params::test::test_replace_params`

To tests/add_foreign_utxo.rs added
- `test_add_planned_psbt_input`

To tests/psbt.rs added
- `test_create_psbt`
- `test_create_psbt_insufficient_funds_error`
- `test_create_psbt_maturity_height`
- `test_create_psbt_cltv`
- `test_create_psbt_csv`
- `test_replace_by_fee_and_recpients`
- `test_create_psbt_utxo_filter`

To tests/wallet.rs added
- `test_spend_non_canonical_txout`

- `test-utils`: Add `insert_tx_anchor` test helper for adding
a transaction to the wallet with associated anchor block.

ValuedMammal committed:
docs: Add `psbt.rs`, `replace_by_fee.rs` to examples

ValuedMammal committed:
docs: Add example code snippets to `examples` directory

- `examples/psbt.rs`
- `examples/replace_by_fee.rs`

ValuedMammal commented:
In the last push

- Removed `PsbtParams::change_policy` member
- Improved `SelectionStrategy::LowestFee` by exposing longterm_feerate, max_rounds
- Added `ChangeScript` support to `PsbtParams`. Since `ChangeScript` is not yet Clone-able, had to add a separate `Wallet::change_script` helper, and update `parse_params` to handle only the assets and indexed txouts

noahjoeris commented on src/wallet/mod.rs:
```diff
@@ -2778,6 +2844,586 @@ impl Wallet {
+            ));
+        let fallback_sequence = params
+            .fallback_sequence
+            .unwrap_or(Sequence::ENABLE_LOCKTIME_NO_RBF);
```
Why default to `ENABLE_LOCKTIME_NO_RBF` here? Before we used `ENABLE_RBF_NO_LOCKTIME`

ValuedMammal replied:
The default was chosen on the assumption that most of the network is supporting `-mempoolfullrbf` (so no explicit signaling required), and in the future signaling BIP125 can make it easier to fingerprint wallets. #66

noahjoeris commented on examples/psbt.rs:
```diff
@@ -0,0 +1,117 @@
+        println!("TxOut: {}", txout.value);
+    }
+
+    let _ = psbt.sign(&xprv, &secp);
```
Should we show the real signing process here by using the wallet instead of hardcoded xpriv?

Something like
```rust
    let external_keymap = wallet.get_signers(External).as_key_map(wallet.secp_ctx());
    let internal_keymap = wallet.get_signers(Internal).as_key_map(wallet.secp_ctx());
    let signer = Signer(external_keymap.into_iter().chain(internal_keymap).collect());

    let _ = psbt
        .sign(&signer, wallet.secp_ctx())
        .map_err(|(_, errors)| anyhow::anyhow!("failed to sign PSBT: {errors:?}"))?;
```

ValuedMammal replied:
I opted to keep the keymaps external to the wallet thinking that `wallet.get_signers()` might be discouraged in the future.

noahjoeris commented on src/wallet/mod.rs:
```diff
@@ -2778,6 +2844,586 @@ impl Wallet {
+        change_script.unwrap_or_else(|| {
+            let change_keychain = self.map_keychain(KeychainKind::Internal);
+            let descriptor = self.public_descriptor(change_keychain);
+            let next_index = self.next_derivation_index(change_keychain);
```
This can reuse the same change script on repeated calls, since this only peeks the next index without revealing it, right? That would be bad for privacy

ValuedMammal replied:
I'll take another look at it. I'm thinking the behavior should be the "next unused address", so if every revealed internal address has been used, we reveal the next one. But until it's used in a transaction and applied to the wallet, it'll be "reused" for the change address. This prevents a runaway in the internal derivation index if many txs are crafted but never go on chain. bitcoindevkit/bdk#1578.

In a multi-keychain world it'd be best to avoid making this decision internally and have the user provide the change script, although it's a bit more hands-on.

ValuedMammal replied:
Done in 3154047a

noahjoeris commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,673 @@
+    pub fn utxos(&self) -> &HashSet<OutPoint> {
+        &self.set
+    }
+
+    /// Remove a UTXO from the currently selected inputs.
+    pub fn remove_utxo(&mut self, outpoint: &OutPoint) -> &mut Self {
+        if self.set.remove(outpoint) {
+            self.utxos.retain(|op| op != outpoint);
+        }
```
Should this also remove matching planned inputs?

noahjoeris commented on src/wallet/mod.rs:
```diff
@@ -2778,6 +2844,586 @@ impl Wallet {
+            return Err(CreatePsbtError::InsufficientFunds(err))?;
+        }
+
+        let original_txs: Vec<OriginalTxStats> = txids_to_replace
```
`build_fee_bump` rejected confirmed originals before bumping and also checked for RBF signaling. Should we reject confirmed txs here too? Otherwise we could return a PSBT that cannot be accepted as an RBF replacement.

ValuedMammal replied:
Yes I think it should reject confirmed txs.

noahjoeris reviewed (commented):
Looking very good overall ✨  I left some small comments

ValuedMammal committed:
fix(wallet): include descendants in RBF OriginalTxStats

Previously, `replace_by_fee_with_rng` only built `OriginalTxStats` from
`txids_to_replace` (the specified conflicts), ignoring their
unconfirmed descendants and undercounting the minimum required fee.

Fix by walking descendants of each txid_to_replace and collecting the
full set into `to_replace`.

`to_replace` is then used in two ways:

1. Filtering UTXOs that must not be selected as new inputs
2. Building the `OriginalTxStats` passed to `RbfParams`

Add `test_replace_by_fee_replaces_descendant_fees` to verify the
behavior. The test constructs a three-tx chain A -> {B, C} and
confirms that replacing A produces a fee >= fee(A) + fee(B) + fee(C).

ValuedMammal committed:
fix: `replace_by_fee` rejects confirmed original txs

If one of the txids-to-replace is already confirmed, then
attempting to replace it will result in an invalid RBF
transaction.

Add `ReplaceByFeeError::TransactionConfirmed` variant to handle
this case. This mirrors the behavior of `build_fee_bump` by
explicitly checking the confirmation status of the original txs
and refusing to build the PSBT if any are confirmed.

ValuedMammal committed:
fix(wallet): Reveal change address on create_psbt/replace_by_fee success

Replace `change_script` helper with `peek_change_info`, which
picks the next unused address without mutating state.

Change methods `create_psbt`, `create_psbt_with_rng`, `replace_by_fee`,
`replace_by_fee_and_recipients`, and `replace_by_fee_with_rng` to
accept `&mut self`.

Once `create_psbt_from_selector` succeeds, check whether the
peeked change SPK appears in the PSBT outputs, and if so, proceed
to `reveal_to_target` and stage the resulting changeset.

Do not call `mark_used` internally to prevent eagerly incrementing the
internal derivation index. The change address is reused on subsequent
calls unless a sync detects the change address as used; Users who need
to prevent this should supply the intended `change_script` directly or
call `mark_used` on the internal address index. This is documented in
a new `# Change address` section on the public `create_psbt` and
`replace_by_fee` docs.

ValuedMammal committed:
fix(params): Remove planned inputs in PsbtParams::remove_utxo

ValuedMammal committed:
docs: Show signing process using bdk_tx::Signer

reez commented on src/wallet/mod.rs:
```diff
@@ -2778,6 +2844,705 @@ impl Wallet {
+        let (assets, txouts) = self.parse_params(&params);
+
+        let PsbtParams {
+            replace: txids_to_replace,
```
Edge case, but should we explicitly guard that at least one selected input overlaps with the inputs of the transactions being replaced? 

The normal path does add those original inputs as `must_spend`, but I think `replace_txs` can be called with an empty set, and `PsbtParams<Rbf>` still allows those UTXOs to be removed via `remove_utxo` before `replace_by_fee` is called. 

I wrote a quick test for that case and it looks like the wallet can still build a PSBT funded by unrelated wallet UTXOs. And even with Full RBF the replacement depends on spending at least one same input, so I think it could be checked directly.

Definitely let me know if I'm missing something though

reez replied:
resolved in https://github.com/bitcoindevkit/bdk_wallet/pull/297#issuecomment-4617634935

reez reviewed (commented):
Overall LGTM, added one comment/question on an edge case

notmandatory commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,706 @@
+    ///
+    /// Panics if `txs` is empty.
+    pub fn replace_txs(self, txs: &[Arc<Transaction>]) -> PsbtParams<Rbf> {
+        assert!(
```
Can we return a result here instead of this assert? Would rather not have a panic caused by user input validation.

notmandatory commented on tests/psbt.rs:
```diff
@@ -1,11 +1,909 @@
+// Test that replacing two unconfirmed txs A, B results in a transaction
+// that spends the inputs of both A and B.
+#[test]
+fn test_replace_by_fee_and_recpients() {
```
nit ⛏️ 
```suggestion
fn test_replace_by_fee_and_recipients() {
```

notmandatory commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,706 @@
+impl PsbtParams<Rbf> {
+    /// Replace spends of the provided `txs`. This will internally set the list of UTXOs
+    /// to be spent.
+    fn replace(&mut self, txs: &[Arc<Transaction>]) {
```
Why does the `txs` param need to be an `&[Arc<Transaction>]` ?  can it be `&[Transaction]` since the code below on 528 is cloning the inner ref and not the `Arc` so there is no additional efficiency.

Same comment for the callers `PsbtParams<>::replace_txs()` and `Wallet::replace_by_fee_and_recipients()`.

notmandatory reviewed (commented):
There's a lot of great work here, I mainly focused my review on the examples, APIs and confirmed no breaking changes, I only have a couple questions and a small nit to fix.  I also ran a couple different LLM models at it and they gave it high marks without finding any significant issues.

ValuedMammal committed:
refactor(psbt): Change replace_txs to take an iterable of Into<Arc<Transaction>>

The empty txs case is now caught in `replace_by_fee_with_rng`,
which returns `ReplaceByFeeError::NoOriginalTransactions` when
`params.replace` is empty.

Adds a test in `tests/psbt.rs` covering the error path.

oleonardolima commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,665 @@
+            self.inputs.push(input);
+        }
+        self
+    }
```
I got a bit confused on the usage of both `self.utxos` and `self.set`, shouldn't the `input` outpoint be addded to `self.utxos` here too ?

nymius replied:
+1

110CodingP replied:
If I understand correctly then, `utxos` corresponds to local UTXOs and `inputs` corresponds to foreign UTXOs in the old TxBuilder and `set` is a hashset of all manually selected utxos (local + foreign) for quick membership checks.

But then do we plan to provide the old invariants? By old invariants I mean,
"Foreign UTXOs are not prioritized over local UTXOs. If a local UTXO is added to the manually selected list, it will replace any conflicting foreign UTXOs. However, a foreign UTXO cannot replace a conflicting local UTXO."

110CodingP commented on examples/psbt.rs:
```diff
@@ -0,0 +1,120 @@
+    // Create PSBT (which also returns the Finalizer).
+    let (mut psbt, finalizer) = wallet.create_psbt(params)?;
+
+    dbg!(&psbt);
```
nit: is this intentional?

oleonardolima commented on src/wallet/mod.rs:
```diff
@@ -2491,6 +2527,36 @@ impl Wallet {
+    ///
+    /// [relevance]: Indexer::is_tx_relevant
+    /// [indexer]: Self::spk_index
+    pub fn insert_tx<T>(&mut self, tx: T)
```
Is this a MUST to go together with `create_psbt` ? I'd move this to another PR instead.

Does it also need a matching (or existing) `WalletEvent` (by using apply_update_events) ?

Also, having the `.expect()` in a public API feels a bit odd.

110CodingP commented on tests/psbt.rs:
```diff
@@ -1,11 +1,909 @@
+        hash: Hash::hash(b"1001"),
+    };
+    insert_tx_anchor(&mut wallet, unconfirmed_tx.clone(), confirm_block);
+    insert_checkpoint(&mut wallet, confirm_block);
```
nit: we don't need this right? since `insert_tx_anchor` calls `insert_checkpoint`.

oleonardolima commented on src/wallet/mod.rs:
```diff
@@ -2778,6 +2844,779 @@ impl Wallet {
+        if params.drain_wallet {
+            selector.select_all();
+        } else {
+            match params.coin_selection {
+                SelectionStrategy::SingleRandomDraw => {
+                    // We should have shuffled candidates earlier, so just select
+                    // until the target is met.
+                    selector
+                        .select_until_target_met()
+                        .map_err(CreatePsbtError::InsufficientFunds)?;
+                }
+                SelectionStrategy::LowestFee {
+                    longterm_feerate,
+                    max_rounds,
+                } => {
+                    selector
+                        .select_with_algorithm(selection_algorithm_lowest_fee_bnb(
+                            longterm_feerate,
+                            max_rounds,
+                        ))
+                        .map_err(CreatePsbtError::Bnb)?;
+                }
```
question: the user does not have the option of using a customized CS algorithm anymore ?

oleonardolima replied:
cc @notmandatory

notmandatory replied:
I would prefer setting the coin selection to any struct that implements `CoinSelectionAlgorithm` so that a custom one can be used. That's how `TxBuilder` currently does it and I don't see a compelling reason to change it.

oleonardolima commented on src/wallet/mod.rs:
```diff
@@ -2778,6 +2844,779 @@ impl Wallet {
+///
+/// [`Height`]: bitcoin::absolute::Height
+/// [`Time`]: bitcoin::absolute::Time
+fn status_from_position(pos: ChainPosition<ConfirmationBlockTime>) -> Option<ConfirmationStatus> {
```
nit: `confirmation_status()` would be just fine too, also it could live in `utils.rs` instead

oleonardolima commented on src/wallet/mod.rs:
```diff
@@ -789,6 +802,19 @@ impl Wallet {
     }
 
+    /// List indexed [`FullTxOut`]s.
+    fn list_indexed_txouts(
```
nit: I'd suggest renaming it to `list_txouts` and reuse it in the already existing `list_output` method.

nymius commented on src/wallet/tx_builder.rs:
What's the feature unlocked by this change?

nymius commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,706 @@
+/// Parameters to create a PSBT.
+// TODO: Can we derive `Clone` for this?
+#[derive(Debug)]
+pub struct PsbtParams<C> {
```
Looking in perspective to the motivation for `bdk-tx`, why this is better than `TxBuilder`? What's the difference?

notmandatory replied:
The two advantages I see are: 
1. `bdk-tx` uses the `rust-miniscript` planning module instead of the hard to use `Policy` approach in TxBuilder.
2. By extracting `bdk-tx` into a new crate it can be used by any projects using `rust-bitcoin` and `rust-miniscript`.

nymius replied:
Thanks Steve, In the change to `bdk-tx`, I initially assigned the same weight to the "code complexity" and lack of "flexibility" issues than to `Policy`, I might have been shortsighted.
Rephrasing my question: how this approach is not going to lead us to the same level of complexity of the old `TxBuilder`? Or have we assumed the complexity is insurmountable? (that's a valid answer too).

evanlinjin replied:
@nymius I guess what you mean by "complexity" is "invalid state representation"?

nymius replied:
No, I meant code complexity: the numerous fields, the setters, the long `create_tx` method and all that.
Or the shape of the code didn't matter that much and the main concern were these invalid state representations.

evanlinjin replied:
Note that some complexity in `create_tx` cannot be avoided right now. I'm assuming it can be simplified once `CanonicalView`, it's extract subgraph method and the `bdk_tx_core` glue lands.

nymius commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,706 @@
+impl PsbtParams<Rbf> {
+    /// Replace spends of the provided `txs`. This will internally set the list of UTXOs
+    /// to be spent.
+    fn replace(&mut self, txs: &[Arc<Transaction>]) {
```
Why we cannot create the `PsbtParams` here, i.e., `fn replace(txs: &[Arc<Transaction>]) -> PsbtParams<Rbf>`? Is there any parameter that we would like to keep from whatever we had before? I've checked the example and we are not reusing parameters from `psbt1`, we should?

nymius replied:
`test_replace_params` repeats the pattern, it doesn't reuse the old `PsbtParams`.

nymius commented on tests/psbt.rs:
```diff
@@ -1,11 +1,909 @@
+}
+
+#[test]
+fn test_create_psbt_maturity_height() {
```
A test with no assertions? Shouldn't we check at least the outpoint used belongs to the coinbase tx?

nymius commented on tests/add_foreign_utxo.rs:
```diff
@@ -323,3 +323,56 @@ fn test_add_foreign_utxo_rejects_wrong_non_witness_utxo_even_with_witness_utxo()
 }
+
+#[test]
+fn test_add_planned_psbt_input() -> anyhow::Result<()> {
```
Is `anyhow::Result<()>` really needed here? It is first introduced by this test. Why not replacing the `?` operators by `.expect("...")`?

nymius reviewed (commented):
I need another round of review yet, but these are the things I've found so far.

ValuedMammal committed:
docs: Fix typo

ValuedMammal committed:
fix(wallet): Add NoRecipients and AllOutputsBelowDust error variants

Adds two `CreatePsbtError` variants to cover distinct failure modes
when building a PSBT:

- `NoRecipients`: no recipients were added, or `drain_wallet` was set
    without an explicit `change_script`.

- `AllOutputsBelowDust`: after coin selection the only output fell
    below dust and was dropped to fees, leaving the transaction with
    zero outputs.

Early-exit guards in `create_psbt_with_rng` and `replace_by_fee_with_rng`
return `NoRecipients`. The post-selection guard in `create_psbt_from_selector`
returns `AllOutputsBelowDust`.

Tests:
- `test_create_psbt_no_recipients_error`
- `test_create_psbt_drain_wallet_change_below_dust_error`
- `test_replace_by_fee_drain_wallet_change_below_dust_error`

nymius commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,706 @@
+#[derive(Debug)]
+pub struct PsbtParams<C> {
+    /// Set of selected UTXO outpoints.
+    pub(crate) set: HashSet<OutPoint>,
+    /// List of UTXO outpoints to spend.
+    pub(crate) utxos: Vec<OutPoint>,
+    /// List of planned transaction [`Input`]s.
```
Why we need both?

nymius commented on tests/psbt.rs:
```diff
@@ -1,11 +1,909 @@
+fn test_create_psbt_cltv_timestamp() {
+    use absolute::LockTime;
+
+    let lock_time = LockTime::from_consensus(1734230218);
```
Here, I would use a constant relative to the unix timepstamp threshold, e.g., `LOCKTIME_TIMESTAMP_FLOOR = 500 millions` then `LockTime::from_consensus(LOCKTIME_TIMESTAMP_FLOOR + 1)`

nymius commented on tests/psbt.rs:
```diff
@@ -1,11 +1,909 @@
+
+    // Locktime greater than required
+    {
+        let new_lock_time = 1772167108;
```
Same here than above comment.

110CodingP commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,706 @@
+            {
+                txids_to_replace.remove(&tx_node.txid);
+            } else {
+                utxos.extend(tx.input.iter().map(|txin| txin.previous_output));
```
This also adds the inputs which were foreign UTXOs to `utxos`  field?
Hacked together a test to check this.
```rust
#[test]
fn test_replace_tx_with_planned_input() {
    use KeychainKind::*;

    let (desc, change_desc) = get_test_wpkh_and_change_desc();
    let mut wallet = Wallet::create(desc, change_desc)
        .network(Network::Regtest)
        .create_wallet_no_persist()
        .unwrap();

    let addr = wallet.reveal_next_address(External).address;

    // Fund the wallet with an unconfirmed output.
    let funding_tx = Transaction {
        input: vec![TxIn {
            previous_output: OutPoint::new(Hash::hash(b"funding_parent"), 0),
            ..Default::default()
        }],
        output: vec![TxOut {
            value: Amount::from_sat(200_000),
            script_pubkey: addr.script_pubkey(),
        }],
        ..new_tx(0)
    };
    let funding_op = OutPoint::new(funding_tx.compute_txid(), 0);
    insert_tx(&mut wallet, funding_tx);

    // Create an unconfirmed tx spending the funded UTXO.
    let recip =
        ScriptBuf::from_hex("5120e8f5c4dc2f5d6a7595e7b108cb063da9c7550312da1e22875d78b9db62b59cd5")
            .unwrap();
    let op2 = OutPoint::new(Hash::hash(b"txid"), 2);
    let txout = TxOut {
        value: Amount::ZERO,
        script_pubkey: ScriptBuf::new_p2a(),
    };
    let psbt_input = bitcoin::psbt::Input {
        witness_utxo: Some(txout),
        ..Default::default()
    };
    let planned_input = bdk_tx::Input::from_psbt_input(
        op2,
        Sequence::ENABLE_LOCKTIME_NO_RBF,
        psbt_input,
        /* satisfaction_weight: */ 0,
        /* status: */ None,
        /* is_coinbase: */ false,
        /* absolute_timelock: */ None,
    ).unwrap();

    let mut params = PsbtParams::default();
    params
        .add_utxos(&[funding_op])
        .add_planned_input(planned_input.clone())
        .add_recipients([(recip, Amount::from_sat(100_000))]);
    let unconfirmed_tx = wallet.create_psbt(params).unwrap().0.unsigned_tx;
    insert_tx(&mut wallet, unconfirmed_tx.clone());
    let params = PsbtParams::default().replace_txs(&[Arc::new(unconfirmed_tx)]);

    // this fails
    let psbt = wallet.replace_by_fee(params).unwrap().0;
}
```

110CodingP commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,706 @@
+    /// to be spent.
+    fn replace(&mut self, txs: &[Arc<Transaction>]) {
+        self.utxos.clear();
+        self.set.clear();
```
Why do we not clear the `inputs`? 

Also clearing the `set` can allow duplicate entries in `inputs`.
```rust
let (desc, change_desc) = get_test_wpkh_and_change_desc();
    let mut wallet = Wallet::create(desc, change_desc)
        .network(Network::Regtest)
        .create_wallet_no_persist()
        .unwrap();

    let addr = wallet.reveal_next_address(External).address;

    // Fund the wallet with an unconfirmed output.
    let funding_tx = Transaction {
        input: vec![TxIn {
            previous_output: OutPoint::new(Hash::hash(b"funding_parent"), 0),
            ..Default::default()
        }],
        output: vec![TxOut {
            value: Amount::from_sat(200_000),
            script_pubkey: addr.script_pubkey(),
        }],
        ..new_tx(0)
    };
    let funding_op = OutPoint::new(funding_tx.compute_txid(), 0);
    insert_tx(&mut wallet, funding_tx);

    // Create an unconfirmed tx spending the funded UTXO.
    let recip =
        ScriptBuf::from_hex("5120e8f5c4dc2f5d6a7595e7b108cb063da9c7550312da1e22875d78b9db62b59cd5")
            .unwrap();
    let op2 = OutPoint::new(Hash::hash(b"txid"), 2);
    let txout = TxOut {
        value: Amount::ZERO,
        script_pubkey: ScriptBuf::new_p2a(),
    };
    let psbt_input = bitcoin::psbt::Input {
        witness_utxo: Some(txout),
        ..Default::default()
    };
    let planned_input = bdk_tx::Input::from_psbt_input(
        op2,
        Sequence::ENABLE_LOCKTIME_NO_RBF,
        psbt_input,
        /* satisfaction_weight: */ 0,
        /* status: */ None,
        /* is_coinbase: */ false,
        /* absolute_timelock: */ None,
    ).unwrap();

    let mut params = PsbtParams::default();
    params
        .add_utxos(&[funding_op])
        .add_recipients([(recip, Amount::from_sat(100_000))]);
    let unconfirmed_tx = wallet.create_psbt(params).unwrap().0.unsigned_tx;
    insert_tx(&mut wallet, unconfirmed_tx.clone());
    let mut params = PsbtParams::default();
    params.add_planned_input(planned_input.clone());
    let mut params = params.replace_txs(&[Arc::new(unconfirmed_tx)]);
    params.add_planned_input(planned_input.clone());

    dbg!(&params); // contains duplicate `Input`s
```

110CodingP commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,706 @@
+    ///
+    /// [`Plan`]: miniscript::plan::Plan
+    /// [`psbt::Input`]: bitcoin::psbt::Input
+    pub fn add_planned_input(&mut self, input: Input) -> &mut Self {
```
Why are we allowed to add a planned_input (using `add_planned_input`) but not a utxo (using `add_utxos`) to a replacement transaction? Is this because we cannot recover the `planned_input`s from the transactions to be replaced?

nymius commented on tests/psbt.rs:
```diff
@@ -1,11 +1,909 @@
+    );
+
+    let addr = wallet.next_unused_address(KeychainKind::External);
+    let rel_locktime = relative::LockTime::from_consensus(6).unwrap();
```
I would use a different value here, as 6 is the default if nothing set, as [line 381 above](https://github.com/ValuedMammal/bdk_wallet/blob/43adb6ea46ebf87fdfc9e5ff078815e8a5d5edc5/tests/psbt.rs#L381) shows

nymius commented on tests/psbt.rs:
```diff
@@ -1,11 +1,909 @@
+            .add_utxos(&[op])
+            .add_recipients([(addr.script_pubkey(), Amount::from_btc(0.42).unwrap())]);
+        let (psbt, _) = wallet.create_psbt(params).unwrap();
+        assert_eq!(psbt.unsigned_tx.input[0].sequence, Sequence(6));
```
I would use a local named constant to point to the place where this value is set.

nymius commented on tests/psbt.rs:
```diff
@@ -1,11 +1,909 @@
+    let fee_c = wallet.calculate_fee(&tx_c).unwrap();
+    insert_tx(&mut wallet, tx_c.clone());
+
+    // The replacement must pay at least the combined fee of all three transactions
+    // (Bitcoin Core RBF Rule 3).
+    let total_original_fee = fee_a + fee_b + fee_c;
+    assert_eq!(total_original_fee.to_sat(), 90_000);
```
Why not checking `replacement_fee >= fee_a + fee_b + fee_c` instead of checking they equal `90_000`? Is there any value on doing this? Could the fee be different than `90_000` after how we have set up the txs above?

nymius commented on tests/psbt.rs:
```diff
@@ -1,11 +1,909 @@
+    params.change_script(change_script);
+    params.drain_wallet();
+    let (psbt, _) = wallet.create_psbt(params).unwrap();
+    assert_eq!(psbt.unsigned_tx.input.len(), 2);
```
We can check the inputs used are `600` and `1000` sats each.

nymius commented on src/wallet/mod.rs:
```diff
@@ -2778,6 +2844,779 @@ impl Wallet {
+        fee_rate: FeeRate,
+        recipients: Vec<(ScriptBuf, Amount)>,
+    ) -> Result<(Psbt, Finalizer), ReplaceByFeeError> {
+        let params = PsbtParams {
+            fee_rate,
+            recipients,
+            ..Default::default()
+        }
+        .replace_txs(txs);
+        self.replace_by_fee_with_rng(params, &mut rand::thread_rng())
```
Is this really an ergonomic improvement?

nymius commented on tests/psbt.rs:
```diff
@@ -1,11 +1,909 @@
+    insert_checkpoint(&mut wallet, confirm_block);
+
+    // Attempting to replace the now-confirmed tx should return TransactionConfirmed.
+    let result = wallet.replace_by_fee_and_recipients(
```
Why not reusing the old params here?

`params.replace(&[Arc::new(unconfirmed_tx)]).fee_rate(fee_rate)`?

This goes in hand with https://github.com/bitcoindevkit/bdk_wallet/pull/297/changes#r3383792188.

nymius commented on tests/wallet.rs:
```diff
@@ -25,6 +25,89 @@ use rand::SeedableRng;
 
+// Test we can select and spend an indexed but not-yet-canonical utxo
+#[test]
+fn test_spend_non_canonical_txout() -> anyhow::Result<()> {
```
I disagree with this, where it's said that unconfirmed txs are non canonical? I would consider this different if there existed a conflicting tx, or even a confirmed conflicting tx, but in this case, the tests is checking unconfirmed tx2 and parents belong to the canonical view.

ValuedMammal committed:
fix(psbt): Preserve planned inputs in `replace`

- Move `add_planned_input` to `impl PsbtParams<CreateTx>`. It was
  previously available on `PsbtParams<C>` (all contexts), allowing
  inputs to be erroneously registered after the state transition.

- Fix `replace()` to preserve pre-registered planned inputs. A `Transaction`
  cannot reconstruct `Input` metadata (psbt fields, satisfaction weight, etc.),
  so planned inputs must be added before calling `replace_txs`. We remove any
  planned input whose `prev_txid` is directly in the replacement set.

- Add `ReplaceByFeeError::ConflictingInput(OutPoint)` in
  `replace_by_fee_with_rng` after computing the full `to_replace` set,
  and validate every manually-selected input against it.

- Add `test_replace_tx_with_planned_input`,
  `test_replace_strips_conflicting_planned_input`, and
  `test_replace_by_fee_conflicting_input_descendant`.

ValuedMammal committed:
fix: Remove unused methods

ValuedMammal commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,843 @@
+        self
+    }
+
+    /// Drain wallet.
+    ///
+    /// This will force selection of the available input candidates. As such, the option is only
+    /// applied to inputs that meet the spending criteria.
+    pub fn drain_wallet(&mut self) -> &mut Self {
```
Perhaps `select_all` would be a better name for describing what it actually does.

evanlinjin commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,843 @@
+    /// Target fee rate.
+    pub(crate) fee_rate: FeeRate,
+    /// Whether to spend all available coins.
+    pub(crate) drain_wallet: bool,
```
Looking at how `drain_wallet` functions, I think `DrainAll` should be a `SelectionStrategy` (or as you suggested, `SelectAll`).

evanlinjin commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,843 @@
+pub struct PsbtParams<C> {
+    /// Set of selected UTXO outpoints, `HashSet` ensures uniqueness
+    pub(crate) set: HashSet<OutPoint>,
+    /// List of UTXO outpoints to spend.
+    pub(crate) utxos: Vec<OutPoint>,
+    /// List of planned transaction [`Input`]s.
+    pub(crate) inputs: Vec<Input>,
```
Instead of having both `utxos` (resolved by wallet) and `inputs` (external UTXOs), I think there should be an unified enum:

```rust
enum InputSpec {
    Owned(OutPoint),   // to be resolved by wallet
    Planned(Input),    // taken as-is
}
```

This will fix an ordering bug with `Untouched` - where the order does not respect the order the UTXOs were introduced if we have both wallet-resolved and external inputs.

evanlinjin reviewed (commented):
Also, what do you think about a staged API? 

```rust
// Stage 1 — wallet resolves a snapshot of spendable inputs
let coins = wallet.candidates()       // -> InputCandidates
    .with_assets(assets)
    .maturity_height(h)
    .filter(|txo| ...)
    .add_foreign(input)               // planned/foreign inputs
    .replace(txs)?;                   // RBF seeds forced inputs here

// Stage 2 — wallet builds the tx from that set
let (psbt, finalizer) = wallet.create_psbt(coins, CreateTxParams)?;
let (psbt, finalizer) = wallet.create_rbf(coins, RbfParams)?;
```

evanlinjin commented on Cargo.toml:
```diff
@@ -21,6 +21,7 @@ unexpected_cfgs = { level = "warn", check-cfg = ['cfg(coverage_nightly)'] }
 
 [dependencies]
 bdk_chain = { version = "0.23.3", features = ["miniscript", "serde"], default-features = false }
+bdk_tx = { version = "0.2.0", default-features = false }
```
Not sure if this is related to this PR, but we should pin bitcoin-units to 0.1.3 or higher. Any lower, and `FeeRate::from_sat_per_vb_u32` would not be available.

evanlinjin commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,843 @@
+#[non_exhaustive]
+pub enum SelectionStrategy {
+    /// Single random draw.
+    #[default]
```
I would argue that `LowestFee` should be the default strategy - as it creates better results.

evanlinjin commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,843 @@
+    ///
+    /// [`UnknownUtxo`]: crate::wallet::error::CreatePsbtError::UnknownUtxo
+    /// [`Wallet::create_psbt`]: crate::Wallet::create_psbt
+    pub fn add_utxos(&mut self, outpoints: &[OutPoint]) -> &mut Self {
```
I think that a consuming builder pattern has better ergonomics. The current (non-consuming) version, forces the caller to instantiate on a separate line and keep a mutable variable around.

```suggestion
    pub fn add_utxos(mut self, outpoints: &[OutPoint]) -> Self {
```

`PsbtParams` will then have `#[must_use]`.

evanlinjin replied:
I guess a non-consuming builder is slightly better API for someone who needs conditional construction. I'll be happy either way.

evanlinjin commented on src/wallet/mod.rs:
```diff
@@ -2778,6 +2804,809 @@ impl Wallet {
+            .iter()
+            .filter_map(|&txid| {
+                let tx = self.tx_graph.graph().get_tx(txid)?;
+                self.calculate_fee(&tx).ok()
```
If the fee cannot be calculated, we may undershoot the RBF fee floor. Maybe we should return an error?

evanlinjin commented on src/psbt/params.rs:
```diff
@@ -0,0 +1,843 @@
+        self
+    }
+
+    /// Set the [`Descriptor`] or raw [`Script`] to be used for generating the change output.
```
We should probably note that if this is not explicitly specified, the wallet will still derive a change address.

evanlinjin commented on src/wallet/error.rs:
```diff
@@ -365,3 +366,112 @@ impl fmt::Display for BuildFeeBumpError {
+    PreviousFee(bdk_chain::tx_graph::CalculateFeeError),
+    /// Original transaction could not be found
+    MissingTransaction(Txid),
+    /// One of the transactions to be replaced is already confirmed
+    TransactionConfirmed(Txid),
```
I don't think this error variant earns it's keep.

* Most callers would want to check whether the transaction is still unconfirmed before trying to replace it. A caller that doesn't do that obviously wants to create a transaction that conflicts with a confirmed one.

* The guard is partial. It only catches "already confirmed at build time". The original transaction can be confirmed before the replacement is ready to broadcast.
